incubator/binding-kafka-grpc/src/main/java/io/aklivity/zilla/runtime/binding/kafka/grpc/internal/stream/KafkaGrpcRemoteServerFactory.java

```diff
@@ -0,0 +1,1994 @@
+    private final SignalFW signalRO = new SignalFW();
+
+    private final WindowFW.Builder windowRW = new WindowFW.Builder();
+    private final ResetFW.Builder resetRW = new ResetFW.Builder();
+
+    private final GrpcBeginExFW grpcBeginExRO = new GrpcBeginExFW();
+    private final GrpcDataExFW grpcDataExRO = new GrpcDataExFW();
+    private final GrpcResetExFW resetExRO = new GrpcResetExFW();
+    private final GrpcAbortExFW abortExRO = new GrpcAbortExFW();
+
+    private final ExtensionFW extensionRO = new ExtensionFW();
+    private final KafkaBeginExFW kafkaBeginExRO = new KafkaBeginExFW();
+    private final KafkaDataExFW kafkaDataExRO = new KafkaDataExFW();
+    private final GrpcQueueMessageFW queueMessageRO = new GrpcQueueMessageFW();
+
+    private final GrpcBeginExFW.Builder grpcBeginExRW = new GrpcBeginExFW.Builder();
+    private final GrpcDataExFW.Builder grpcDataExRW = new GrpcDataExFW.Builder();
+    private final GrpcResetExFW.Builder grpcResetExRW = new GrpcResetExFW.Builder();
+    private final GrpcAbortExFW.Builder grpcAbortExRW = new GrpcAbortExFW.Builder();
+
+    private final KafkaBeginExFW.Builder kafkaBeginExRW = new KafkaBeginExFW.Builder();
+    private final KafkaDataExFW.Builder kafkaDataExRW = new KafkaDataExFW.Builder();
+    private final GrpcQueueMessageFW.Builder queueMessageRW = new GrpcQueueMessageFW.Builder();
+
+    private final Long2ObjectHashMap<KafkaGrpcBindingConfig> bindings;
+    private final BufferPool bufferPool;
+    private final MutableDirectBuffer writeBuffer;
+    private final MutableDirectBuffer extBuffer;
+    private final BindingHandler streamFactory;
+    private final LongUnaryOperator supplyInitialId;
+    private final LongUnaryOperator supplyReplyId;
+    private final LongSupplier supplyTraceId;
+    private final Signaler signaler;
+    private final int grpcTypeId;
+    private final int kafkaTypeId;
+    private long reconnectAt = NO_CANCEL_ID;
+    private long lastBindingId;
+
+
+    public KafkaGrpcRemoteServerFactory(
+        KafkaGrpcConfiguration config,
+        EngineContext context)
+    {
+        this.bufferPool = context.bufferPool();
+        this.writeBuffer = context.writeBuffer();
+        this.extBuffer = new UnsafeBuffer(new byte[context.writeBuffer().capacity()]);
+        this.streamFactory = context.streamFactory();
+        this.supplyInitialId = context::supplyInitialId;
+        this.supplyReplyId = context::supplyReplyId;
+        this.signaler = context.signaler();
+        this.supplyTraceId = context::supplyTraceId;
+        this.bindings = new Long2ObjectHashMap<>();
+        this.grpcTypeId = context.supplyTypeId(GRPC_TYPE_NAME);
+        this.kafkaTypeId = context.supplyTypeId(KAFKA_TYPE_NAME);
+    }
+
+    @Override
+    public void attach(
+        BindingConfig binding)
+    {
+        KafkaGrpcBindingConfig newBinding = new KafkaGrpcBindingConfig(binding);
+        lastBindingId = binding.id;
```

Comment:
I really like the approach of using `Signaler` to initiate the remote server streams. 👍 

However, I think this concept of retaining the `lastBindingId` will not work if there are multiple `kafka-grpc` bindings because the `lastBindingId` would be overwritten during subsequent call to `attach(...)` before the first signal is processed.

Instead I suggest we create the list of `RemoteServers` here, based on `binding.routes`, and upon receiving the signal, notify each one to `.initiate()`. 🙂